lightning/src/ln/channelmanager.rs

```diff
@@ -3011,45 +3011,62 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+														short_channel_id: prev_short_channel_id,
+														outpoint: prev_funding_outpoint,
+														htlc_id: prev_htlc_id,
+														incoming_packet_shared_secret: incoming_shared_secret,
+													})
+												}
+											}
 											macro_rules! fail_forward {
 												($msg: expr, $err_code: expr, $err_data: expr) => {
 													{
 														log_info!(self.logger, "Failed to accept/forward incoming HTLC: {}", $msg);
-														let htlc_source = HTLCSource::PreviousHopData(HTLCPreviousHopData {
-															short_channel_id: short_chan_id,
-															outpoint: prev_funding_outpoint,
-															htlc_id: prev_htlc_id,
-															incoming_packet_shared_secret: incoming_shared_secret,
-														});
-														failed_forwards.push((htlc_source, payment_hash,
+														failed_forwards.push((get_failure_htlc_source!(), payment_hash,
 																HTLCFailReason::Reason { failure_code: $err_code, data: $err_data }
 														));
 														continue;
 													}
 												}
 											}
+											macro_rules! fail_phantom_forward {
+												($msg: expr, $err_code: expr, $err_data: expr, $phantom_shared_secret: expr) => {
+													{
+														log_info!(self.logger, "Failed to accept/forward incoming phantom node HTLC: {}", $msg);
+														let packet = onion_utils::build_failure_packet(&$phantom_shared_secret, $err_code, &$err_data[..]).encode();
+														let error_data =  onion_utils::encrypt_failure_packet(&$phantom_shared_secret, &packet);
+														failed_forwards.push((get_failure_htlc_source!(), payment_hash,
+																HTLCFailReason::LightningError { err: error_data }
+														));
+														continue;
+													}
+												}
+											}
 											if let PendingHTLCRouting::Forward { onion_packet, .. } = routing {
 												let phantom_secret_res = self.keys_manager.get_node_secret(Recipient::PhantomNode);
 												if phantom_secret_res.is_ok() && fake_scid::is_valid_phantom(&self.fake_scid_rand_bytes, short_chan_id) {
-													let shared_secret = {
+													let phantom_shared_secret = {
 														let mut arr = [0; 32];
 														arr.copy_from_slice(&SharedSecret::new(&onion_packet.public_key.unwrap(), &phantom_secret_res.unwrap())[..]);
 														arr
 													};
-													let next_hop = match onion_utils::decode_next_hop(shared_secret, &onion_packet.hop_data, onion_packet.hmac, payment_hash) {
+													let next_hop = match onion_utils::decode_next_hop(phantom_shared_secret, &onion_packet.hop_data, onion_packet.hmac, payment_hash) {
 														Ok(res) => res,
 														Err(onion_utils::OnionDecodeErr::Malformed { err_msg, err_code }) => {
-															fail_forward!(err_msg, err_code, Vec::new());
+															fail_phantom_forward!(err_msg, err_code, Vec::new(), phantom_shared_secret);
```

Comment:
Hmm, shouldn't malformed not get encrypted by the phantom node? Malformed generally is "send it back to previous node and let them encrypt it".

Reply:
Oops, will fix